src/cholesky/lkjcholesky.jl

```diff
@@ -0,0 +1,183 @@
+            end
+            isapprox(s, 1) || return false
+        end
+    else  # R.uplo === 'L'
+        for i in 1:p
+            s = zero(T)
+            for j in 1:i
+                s += factors[iinds[i], jinds[j]]^2
+            end
+            isapprox(s, 1) || return false
+        end
+    end
+    return true
+end
+
+function mode(d::LKJCholesky)
+    p = dim(d)
+    factors = Matrix{partype(d)}(I, p, p)
+    return Cholesky(factors, d.uplo, 0)
+end
+
+params(d::LKJCholesky) = (d.d, d.η, d.uplo)
+
+@inline partype(::LKJCholesky{T}) where {T <: Real} = T
+
+#  -----------------------------------------------------------------------------
+#  Evaluation
+#  -----------------------------------------------------------------------------
+
+function logkernel(d::LKJCholesky, R::Cholesky)
+    factors = R.factors
+    p, η = params(d)
+    c = p + 2(η - 1)
+    T = typeof(one(c) * log(one(eltype(factors))))
+    logp = zero(T)
+    di = diagind(factors)
+    for i in 2:p
+        logp += (c - i) * log(factors[di[i]])
+    end
+    return logp
+end
+
+logpdf(d::LKJCholesky, R::Cholesky) = logkernel(d, R) + d.logc0
+
+pdf(d::LKJCholesky, R::Cholesky) = exp(logpdf(d, R))
+
+loglikelihood(d::LKJCholesky, R::Cholesky) = logpdf(d, R)
+function loglikelihood(d::LKJCholesky, Rs::AbstractArray{<:Cholesky})
+    return sum(R -> logpdf(d, R), Rs)
+end
+
+#  -----------------------------------------------------------------------------
+#  Sampling
+#  -----------------------------------------------------------------------------
+
+function rand(rng::AbstractRNG, d::LKJCholesky)
+    p = dim(d)
+    T = eltype(d)
+    factors = Matrix{T}(undef, p, p)
+    R = Cholesky(factors, d.uplo, 0)
+    return rand!(rng, d, R)
+end
+function rand(rng::AbstractRNG, d::LKJCholesky, dims::Dims)
+    p = dim(d)
+    uplo = d.uplo
+    T = eltype(d)
+    TChol = Cholesky{T,Matrix{T}}
+    Rs = Array{TChol}(undef, dims)
+    for i in eachindex(Rs)
+        factors = Matrix{T}(undef, p, p)
+        Rs[i] = R = Cholesky(factors, uplo, 0)
+        rand!(rng, d, R)
+    end
+    return Rs
+end
+
+rand!(rng::AbstractRNG, d::LKJCholesky, R::Cholesky) = _lkj_cholesky_vine_sampler!(rng, d, R)
+
+function _lkj_cholesky_vine_sampler!(rng::AbstractRNG, d::LKJCholesky, R::Cholesky)
+    p, η = params(d)
+    factors = R.factors
+    if R.uplo === 'U'
+        z = _lkj_vine_rand_cpcs!(transpose(factors), p, η, rng)
+        _cpcs_to_cholesky!(factors, z, p)
+    else  # uplo === 'L'
+        z = _lkj_vine_rand_cpcs!(factors, p, η, rng)
+        _cpcs_to_cholesky!(transpose(factors), z, p)
+    end
+    return R
+end
+
+# sample partial canonical correlations z using the vine method from Section 2.4 in LKJ (2009 JMA)
+function _lkj_vine_rand_cpcs!(z, d::Integer, η::Real, rng::AbstractRNG)
+    T = eltype(z)
+    β = η + T(d - 1) / 2
+    @inbounds for i in 1:(d - 1)
+        β -= T(0.5)
+        z[i, (i + 1):d] .= 2 .* rand.(rng, Ref(Beta(β, β))) .- 1
```

Comment:
I would assume that this is less performant than just iterating with a for loop and a precomputed sampler since involves the heavy broadcasting machinery and recomputes the values for the Beta sampler multiple times:
```suggestion
        spl = sampler(Beta(β, β))
        for j in (i + 1):d
            z[i, j] = 2 * rand(rng, spl) - 1
        end
```
I guess if one would iterate in the natural order (rows at the inner level) than one could not use the sampler?

Reply:
I'd expect so, but it seems that the broadcasting version is 3x faster than with this change. Perhaps this is because `sampler(Beta(β, β))` is not type-stable.

Reply:
Ah yeah, maybe one has to add a function barrier outside of the for loop (similar to the suggestion in https://github.com/JuliaStats/Distributions.jl/pull/1281#issuecomment-836246397), i.e.,
```julia
function _loop!(rng, z, i, js, sampler)
    @inbounds for j in js
        z[i, j] = 2 * rand(rng, sampler) - 1
    end
end
```
and then use
```julia
    _loop!(rng, z, i, (i + 1):d, sampler(Beta(β, β)))
```
in the for loop.

Reply:
It's still slightly slower for small matrices but faster for large matrices, so this is worth it.